### PR TITLE
feat(components): Add support for load more trigger to combobox

### PR DESCRIPTION
**Pull request by mattander**
## Motivations

Some clients have huge lists of tags and we need a way to keep those queries paged but load more as the SP reaches the bottom of the scroll list. This provides a low-touch option to just drop a custom component into the bottom of the list, offloading the business logic for loading more and showing loading indicators etc to the implementer. 

## Changes

- add `listEndEnhancer` support to Combobox

## Testing

Check the new infinite scroll story for an example of how this can be implemented. 

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

**Review by darryltec — CHANGES_REQUESTED** [review]
First, did you talk to any Atlantis folks about this before you wrote any code?

Second, thought about this more and as cool as having a low touch point and off-loading responsibility to the implementer/consumer, Atlantis isn't a generic design system.

> What ever do you mean?

Imagine a spectrum that has generic and opinionated on opposite ends.

We can classify Material, Ant, Vuetify, or any headless component systems as "Generic". It can work anywhere with any design. It works really well if everyone, and I mean everyone, in the company have a mutual understanding on how to use it. It, unfortunately, comes at a cost of consistency since it is as easy to get into an inconsistent implementation and design as it is to use it.

Polaris, Atlassian, or any DS that is created by a B2B SaaS company (that isn't Google) is on the opinionated side of things. You can only do certain things as they rely on their system to ensure their design and code are consistent. Not much configurability, but that's by design (figuratively and literally). They have certain patterns that isn't negotiable.

image

Atlantis is closer to opinionated. We value _consistency over configurability_ because we don't want you and the designer to rethink mundane stuff like loading all the time. That should be built in the component.

Now, we also favour _context over consistency_ as being consistent doesn't translate well when the context requires a different approach.

### So, what am I getting at?

In the essence of speed and time, I'm just gonna be straight.

Consistency-wise, if you look at how we've done loading more when you've hit the end (DataList, Chips) and other systems Like [Polaris](https://polaris.shopify.com/components/selection-and-input/combobox) (Opinionated) or [React Native](https://polaris.shopify.com/components/selection-and-input/combobox) (Generic) all of them just uses a simple callback and a loading prop.

Context-wise, this is a pretty known pattern and something that no designers nor dev should have to think about. At the end of the day, you'd just want it to work with minimal setup.

So, the implementation should just be an `onLoadMore` callback and update the `loading` prop to add a spinner (consistency) at the end of the list when it is not empty.

### With that said...

I'm not saying we won't do something like this approach. It really depends on the context and how much that design changes. DataList and RadioOption is a good example where you are allowed to add as much customizability as you want. But only on ones that we know needs it and isn't a known pattern.

So if you think your approach is good, not just for the devs who prefers customization, but also for the devs and design who just want something that works. I'd like to have a more compelling argument as to why we should do it the way you did it.

To recap, Atlantis isn't a design system for all, it is a design system for Jobber.

LMK if you want to chat face-to-face about this 😁

**Comment by mattander** [comment]
@darryltec Thanks for the detailed feedback.  I definitely understand the points you've made here and am happy to adjust the approach to be more similar to what your team has done in `DataList`.

I've pushed up a commit with the majority of the changes. I've got a couple questions about finishing touches:

1. How would you like to handle the loading states in combination with the custom search cases. Right now that test case is failing because I'm not yet certain how we want all the various states to play out. List has options but is loading, show spinner instead of glimmer? I'm just having trouble figuring out how to handle the difference between loading a search response and loading more options because loading more has a spinner at the end of the list whereas a user performing a search will see the top of the list. Any guidance here would be appreciated.
2. I've structured the load more trigger to be generic since there's nothing it does that really needs to be specific to combobox. I've mostly copy/pasted the datalist approach. I think this could be a handy generic component that could get used in datalist too to avoid writing the same in view trigger a bunch. I made it so it can use `children` for a custom load state indicator if it's provided but fallback on a small spinner. If you'd prefer me to make this more specific to combobox, I can modify it. On the flipside, if you like the generic approach, I'm happy to make that a separate component.

Thanks!

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.tsx [hunk 1]
```diff
@@ -51,6 +52,16 @@ export function ComboboxContentList(props: ComboboxListProps): JSX.Element {
...
+          {props.onLoadMore && (
+            <ComboboxLoadMore
+              onLoadMore={props.onLoadMore}
+              loading={props.loading}
+            >
+              <div className={styles.loadingContainer}>
+                <Spinner size="small" />
+              </div>
+            </ComboboxLoadMore>
+          )}
```
Comment (darryltec) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.tsx:64]: So this implementation couples the loading state to the `onLoadMore`. I get it tho, it does feel like there's 3 states for the loading state.

But there's actually only 2. 
- If it's loading and there are no options (Glimmer)<br/>image
- If it's loading and there are options (Spinner stuck to the bottom and always visible)<br/>image

Both scenarios don't need to know if you have an `onLoadMore` (or `onSearch` if that doesn't work as expected (it probably isn't looking at where it is in the code 😆))

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMoreTrigger.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+}) {
+  const [inViewRef, isInView] = useInView<HTMLDivElement>();
+
+  useEffect(() => {
+    isInView && onLoadMore?.();
+  }, [isInView]);
+
+  return (
+    <div
+      data-testid="ATL-Loadmore-Trigger"
```
Comment (darryltec) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMoreTrigger.tsx:18]: You should narrow this down to the name of this component. That's too generic.

**Review by darryltec — DISMISSED** [review]
This is close! You may encounter some weirdness with the onSearch loading state but I'll leave it up to you if you want to improve it or just do what you came here to do 😆

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  readonly onLoadMore: () => void;
+  readonly loading: boolean | undefined;
+}
+
+export function ComboboxLoadMore({
+  children,
+  loading,
+  onLoadMore,
+}: ComboboxLoadMoreProps) {
+  if (loading) return children ? <>{children}</> : <Spinner size="small" />;
```
[the remark this answers is not in the document]
Comment (mattander) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.tsx:15]: this was mostly here for if we wanted to make this into a generic trigger. I think we'll leave that aside for the sake of time and it wouldn't be hard to do another time as a tech investment.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMoreTrigger.tsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import React, { useEffect } from "react";
+import { useInView } from "@jobber/hooks/useInView";
+import styles from "./ComboboxLoadMore.css";
+
+export function ComboboxLoadMoreTrigger({
+  onLoadMore,
+}: {
+  readonly onLoadMore: () => void;
+}) {
```
[the remark this answers is not in the document]
Comment (mattander) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMoreTrigger.tsx:9]: thanks! missed that one.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.css [hunk 5]
```diff
@@ -0,0 +1,4 @@
+.trigger {
+  position: relative;
```
[the remark this answers is not in the document]
Comment (mattander) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.css:2]: took it straight outta `DataListLoadMoreTrigger`

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (darryltec) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.tsx:15]: i'd add it when we need it. Much easier to add it than to support something that may or may not be needed.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.css (hunk 5 above)
[the remark this answers is not in the document]
Comment (mattander) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.css:2]: but yeah, now that you called it out, this is pointless. there's no child elements lol. i'll get rid of it. this was just copy pasta.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.tsx (hunk 1 above)
Comment (mattander) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.tsx:64]: Perfect. Love it.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.css [hunk 6]
```diff
@@ -61,3 +62,15 @@
 .loadingContainer {
   padding: var(--space-base) var(--space-base);
 }
+
+.loadingContainer.hasOptions {
+  display: flex;
+  justify-content: center;
+
+  position: fixed;
```
Comment (MichaelParadis) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.css:70]: Minor this, I noticed that the spinner renders weirdly with this position fixed here. It seems to be resolved if this is removed (see video)


https://github.com/GetJobber/atlantis/assets/17303106/f660480c-14d9-40bd-a717-4243c0c6fa65

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.tsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+interface ComboboxLoadMoreProps extends React.PropsWithChildren {
+  readonly onLoadMore: () => void;
+}
+
+export function ComboboxLoadMore({ onLoadMore }: ComboboxLoadMoreProps) {
+  const [inViewRef, isInView] = useInView<HTMLDivElement>();
+
+  useEffect(() => {
+    isInView && onLoadMore?.();
```
Comment (MichaelParadis) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.tsx:13]: Nit: `onLoadMore` is required so it won't ever be undefined

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import React, { useEffect } from "react";
+import { useInView } from "@jobber/hooks/useInView";
+import styles from "./ComboboxLoadMore.css";
+
+interface ComboboxLoadMoreProps extends React.PropsWithChildren {
```
Comment (MichaelParadis) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxLoadMore/ComboboxLoadMore.tsx:5]: Nit: There aren't any children used so I don't think `extends ...` is needed

File: docs/components/Combobox/Web.stories.tsx [hunk 9]
```diff
@@ -332,6 +404,9 @@ DynamicAction.args = {};
 export const CustomSearch = ComboboxCustomSearch.bind({});
 CustomSearch.args = {};
 
+export const InfiniteScroll = ComboboxInfiniteScroll.bind({});
+CustomSearch.args = {};
```
Comment (MichaelParadis) [inline docs/components/Combobox/Web.stories.tsx:408]:
```suggestion
InfiniteScroll.args = {};
```

**Review by MichaelParadis — CHANGES_REQUESTED** [review]
Super minor changes. Mainly some CSS and minor types to change. Let me know if you have any questions. Thanks for working on this

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.css (hunk 6 above)
Comment (MichaelParadis) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.css:70]: Removing the `position: fixed` and `bottom: 0` seems to resolve this

Comment (MichaelParadis) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.css:70]: @mattander feel free to ignore this. I just saw that :face-palm:

File: docs/components/Combobox/Web.stories.tsx (hunk 9 above)
Comment (mattander) [reply docs/components/Combobox/Web.stories.tsx:408]: Thanks. Missed that.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.css (hunk 6 above)
Comment (mattander) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxContentList/ComboboxContentList.css:70]: Hah okay! Thanks!

**Review by MichaelParadis — APPROVED** [review]
LGTM! Once CI finishes I will merge and tag you

**Comment by MichaelParadis** [comment]
@mattander I am merging right now. A new npm version will be published shortly after the merge
